File: imp/objective-c/src/fit/ColumnFixture.m
//
//  ColumnFixture.m
//  FIT
//
//  Created by Steven Newton on Mon Sep 29 2003.
//  Copyright (c) 2002 Cunningham & Cunningham, Inc.
//  Released under the terms of the GNU General Public License version 2 or later.
//

#import "ColumnFixture.h"


@implementation ColumnFixture

- (void) doRows: (Parse *)rows {
    [self bind: [rows parts]];
    [super doRows: [rows more]];
}

- (void) doRow: (Parse *)row {
    hasExecuted = NO;
NS_DURING
    [self reset];
    [super doRow: row];
    if (!hasExecuted) {
        [self execute];
    }
NS_HANDLER
    [self exception: [row leaf] e: localException];
NS_ENDHANDLER
}

- (void) doCell: (Parse *) cell columnNumber: (int) columnNumber {
    TypeAdapter *a = [columnBindings objectAtIndex: columnNumber];
    NS_DURING
        NSString *text = [cell text];
        if ([text isEqualToString: @""]) {
            [self check: cell adapter: a];
        } else if (a == nil) {
            [self ignore: cell];
            // if this is a field column, set it
        } else if (a->field != nil) {
            [a set: [a parse: text]];
            // else this is a method column, call the method
        } else if (a->method != nil) {
            [self check: cell adapter: a];
        }
    NS_HANDLER
        [self exception: cell e: localException];
    NS_ENDHANDLER
}

- (void) check: (Parse *) cell adapter: (TypeAdapter *) a {
    [super check: cell adapter: a];
}

- (void) bind: (Parse *)heads {
    int i;
    columnBindings = [NSMutableArray arrayWithCapacity: [heads size]];
    NSString *suffix = @"()";
    for (i = 0; heads != nil; i++, heads = [heads more]) {
        NSString *name = [heads text];
        if (name == nil) {
            return;
        } else if ([name isEqualToString: @""]) {
            [columnBindings insertObject: [NSNull null] atIndex: i];
        } else if ([name hasSuffix: suffix]) {
            name = [name substringToIndex: [name length] - [suffix length]];
            [columnBindings insertObject: [self bindMethod: name] atIndex: i];
        } else {
            [columnBindings insertObject: [self bindField: name] atIndex: i];
        }
            
    }
}

- (TypeAdapter *) bindField: (NSString *) name {
    return [TypeAdapter on: self field: name];
}

- (TypeAdapter *) bindMethod: (NSString *) name {
    return [TypeAdapter on: self method: name];
}

- (void) reset {
    // about to process first cell of row
}

- (void) execute {
    // about to process first method call of row
}

@end

File: imp/ruby/lib/fit/column_fixture.rb
# Copyright (c) 2002 Cunningham & Cunningham, Inc.
# Released under the terms of the GNU General Public License version 2 or later.

require 'fit/fixture'
require 'fit/type_adapter'

module Fit

  class ColumnFixture < Fixture

    attr_accessor :column_bindings, :has_executed
    protected :column_bindings, :column_bindings=, :has_executed, :has_executed=

    # Traversal

    def do_rows rows
      bind rows.parts
      super(rows.more)
    end

    def do_row row
      @has_executed = false
      begin
        reset
        super
        execute if not @has_executed
      rescue Exception => e
        exception row.leaf, e
      end
    end

    def do_cell cell, column_index
      adapter = @column_bindings[column_index]
      begin
        text = cell.text
        if text.empty?
          check cell, adapter
        elsif adapter.nil?
          ignore cell
        elsif adapter.is_output?
          check cell, adapter
        else
          adapter.set(adapter.parse(text))
        end
      rescue Exception => e
        exception cell, e
      end
    end

    def check cell, adapter
      unless @has_executed
        begin
          execute
        rescue Exception => e
          exception cell, e
        end
        @has_executed = true
      end
      super
    end

    def reset
      # about to process first cell of row
    end

    def execute
      # about to process first method call of row
    end

    # Utility

    protected

    def bind heads
      @column_bindings = []
      until heads.nil?
        name = heads.text
        begin
          if name.empty?
            @column_bindings << nil
          elsif name =~ /\(\)$/
            @column_bindings << TypeAdapter.for(self, name[0..-3], true)
          else
            @column_bindings << TypeAdapter.for(self, name, false)
          end
        rescue Exception => e
          exception heads, e
        end
      heads = heads.more
      end
    end

  end

end
